### ISSUE TITLE
Pressing the edge of the touchpad is incorrectly reported as touching it

### ISSUE THREAD

**Issue by LibreVR** [ValveSoftware/openvr#181]
When you are touching the edge of the touchpad it doesn't detect that as touching it, because it wouldn't be able to determine the position of the finger. But if you proceed to press the edge so you trigger the touchpad, it will also incorrectly report that you are touching it in the middle (0,0).

This can cause applications to incorrectly assume you pressed the touchpad in the middle, while in reality you pressed the edge. This can lead to an undesired action.

I suggest changing the behaviour so that pressing the touchpad does not automatically imply you are touching it. It should only report that you are touching the touchpad if it can actually determine the position of the finger.

**Comment by jrbudda** [comment]
Hah, yes I just ran this exact thing.  I worked around it by specifically ignoring presses if the position was exactly 0,0. I actually wasn't checking for buttonTouched in this case but if that's true then this really should be fixed.

**Comment by LibreVR** [comment]
Problem with that approach is that users may actually be pressing the touchpad at 0,0.

**Comment by jrbudda** [comment]
I agree totally, was not suggesting my workaround negates it needing fixed.

**Comment by LibreVR** [comment]
Also you can clearly see the result of this behavior on the render models of the Vive controllers. If you touch the edge of the controller you don't see any dot. But if you press the edge then the dot that indicates your finger position shows up in the middle of the touchpad.

**Comment by LibreVR** [comment]
Also, I think we should hurry with a proper fix before developers start implementing all kinds of workarounds which end up making the behavior of the touchpad inconsistent.

**Comment by chaoticbob** [comment]
Do you have any idea on similar/different it is from this issue? https://github.com/ValveSoftware/openvr/issues/82

**Comment by LibreVR** [comment]
Issue #82 does indeed mention this issue as a sub-problem:

> Once pressed, another problem happens. The touchpad bit will always be set even if untouched. This is reproduceable in the waiting room shipped with SteamVR.

**Comment by JeffLeinbaugh** [comment]
The trackpad behavior was changed in today's beta (1470702757).
http://steamcommunity.com/games/250820/announcements/detail/610621429037542937

Previously, pressing the trackpad button would set both buttonPressed and buttonTouched, even if there was no valid coordinate for the touch (causing this issue). Now, the "pressed" state is not asserted unless there is also a valid touch. "Pressed but not touched" is still not a valid state. As a result, pressing the very edge of the trackpad, where the finger position is unknown, will have no effect (buttonPressed = buttonTouched = 0).

I believe this solution will have the lowest impact on existing applications, but we will be listening to feedback and looking for cases that do break while it's still in beta.

**Comment by LibreVR** [comment]
@JeffLeinbaugh That might've worked if this state was just an edge case, however not touching the pad while being able to press the button is actually very common. And quite a few games do not significantly use the touch position anyway and would be negatively impacted by that change.

I think that most current games will only check the "pressed" state and any change in the "touched" state would not significantly affect them. Adding the functionality to report the last known position of the finger would improve those games.

If you're still worried about breaking current games though, then I think upgrading the interface version is a better way to handle that.

**Comment by thomasbochtler** [comment]
I think the main problem here is that touching and pressing the edges is not a valid state. Why can the coordinates of touches and presses on the edge not be detected correctly? Is this a hardware problem? 

From a developers point of view I would love to the get the pressed button states with valid coordinates, even if the user presses the edge of the touchpad.

**Comment by LibreVR** [comment]
@thomasbochtler It is definitely a hardware problem, the sensor doesn't reach to the edges of the touchpad. Unfortunately the OpenVR drivers aggravate this problem: In the stable version of SteamVR an invalid position will be reported and in the new beta it won't report a pressed state at all.
